engine/runner/src/main/java/org/enso/runner/JavaFinder.java

```diff
@@ -0,0 +1,91 @@
+package org.enso.runner;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Comparator;
+import org.enso.distribution.DistributionManager;
+import org.enso.distribution.Environment;
+import org.enso.runtimeversionmanager.components.GraalRuntime;
+import org.enso.runtimeversionmanager.components.GraalVMVersion;
+import org.enso.runtimeversionmanager.components.GraalVersionManager;
+import org.enso.version.BuildVersion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility class that tries to find installed JDK on the system. */
+final class JavaFinder {
+  private static final Logger logger = LoggerFactory.getLogger(JavaFinder.class);
+
+  private JavaFinder() {}
+
+  /**
+   * Tries to find {@code java} executable on the system. If a system-wide JDK is not found, tries
+   * to find it in the {@link DistributionManager distribution} runtimes.
+   *
+   * @return null if cannot be found. Otherwise, returns the absolute path to the executable, or
+   *     simply {@code java} if it is on the {@code PATH}.
+   */
+  static String findJavaExecutable() {
+    var javaInRuntime = findJavaExecutableInDistributionRuntimes();
+    if (javaInRuntime != null) {
+      return javaInRuntime.toAbsolutePath().toString();
+    }
+    logger.warn("No appropriate JDK found in the distribution runtimes. Trying system-wide JDK.");
+    var javaHome = System.getenv("JAVA_HOME");
+    if (javaHome != null) {
+      var java = new File(javaHome, "bin/java").getAbsoluteFile();
+      if (java.exists()) {
+        return java.getAbsolutePath();
+      }
+    }
+    if (isJavaOnPath()) {
+      return "java";
```

Comment:
Does that work on Windows?

Reply:
I don't expect it to work on windows. This is a fallback that is currently only usable for devs and not anyone else. For all the other use-cases, the java exe should be found in the installed distribution runtimes.